eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToList(inputArray) { 
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let currentIndex = inputArray.length-1;
+
+    function innerArrayToList(inputArray, list = null) {
+        if (currentIndex < 0) {
+            return list;
+        }
+
+        let arrayElement = inputArray[currentIndex];
+        currentIndex--;
+        return innerArrayToList(inputArray, { value: arrayElement, rest: list});
+    }
+
+    return innerArrayToList(inputArray);
+}
+
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToListIterative(inputArray) {
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let listAcc = null;
+
+    for (let i = inputArray.length - 1; i >= 0; i--) {
+        let element = inputArray[i];
+        listAcc = {
+            value: element,
+            rest:listAcc
+        };
+    }
+
+    return listAcc;
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArray(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    let arrayAccumulator = [];
+
+    function innerListToArray(list) {
+        arrayAccumulator.push(list.value);
+        if (list.rest) {
+            innerListToArray(list.rest)
+        }
+        return arrayAccumulator;
+    }
+
+    return innerListToArray(inputList);
```

Comment:
The pattern of closing over arrayAccumulator but then returning it from the function is weird. I'd either not return it, call innerListToArray before the return, and then return it, or I'd make arrayAccumulator an argument.